src/main/java/io/sentry/UploadSourceBundleMojo.java

```diff
@@ -78,79 +82,133 @@ public void execute() throws MojoExecutionException {
     final @NotNull SentryCliRunner cliRunner =
         new SentryCliRunner(
             debugSentryCli, sentryCliExecutablePath, mavenProject, mavenSession, pluginManager);
+
     createDebugMetaPropertiesFile(bundleId);
-    bundleSources(cliRunner, bundleId, sourceBundleTargetDir);
+    collectSources(bundleId, collectedSourcesTargetDir);
+    bundleSources(cliRunner, bundleId, collectedSourcesTargetDir, sourceBundleTargetDir);
     uploadSourceBundle(cliRunner, sourceBundleTargetDir);
   }
 
+  private void collectSources(@NotNull String bundleId, @NotNull File outputDir) {
+    final @Nullable ISpan span = SentryTelemetryService.getInstance().startTask("collectSources");
+    logger.debug("Collecting files from source directories");
+
+    if (!outputDir.exists()) {
+      outputDir.mkdirs();
+    }
+
+    final @Nullable List<String> sourceRootsPaths = mavenProject.getCompileSourceRoots();
+    final @NotNull List<String> sourceDirsPaths =
+        sourceRootsPaths == null ? new ArrayList<>() : new ArrayList<>(sourceRootsPaths);
+    sourceDirsPaths.removeIf(Objects::isNull);
+    sourceDirsPaths.addAll(additionalSourceDirsForSourceContext);
+
+    if (sourceDirsPaths.isEmpty()) {
+      logger.info("No source directories to collect and bundle");
+    } else {
+      logger.debug(
+          "Copying files from the following directories to {} for inclusion in source bundle: {}",
+          outputDir,
+          String.join(",", sourceDirsPaths));
+    }
+    int count = 0;
+    for (final @NotNull String sourceDirPath : sourceDirsPaths) {
+      try {
+        final @NotNull File sourceDir = new File(sourceDirPath);
+        final @NotNull Path sourceDirAbsolutePath = sourceDir.toPath().toAbsolutePath().normalize();
+
+        if (!sourceDir.exists()) {
+          logger.error(
+              "Not collecting sources in {}: directory does not exist", sourceDirAbsolutePath);
+          continue;
+        }
+        if (!sourceDir.isDirectory()) {
+          logger.error("Not collecting sources in {}: not a directory", sourceDirAbsolutePath);
+          continue;
+        }
+        logger.debug("Collecting sources in {}", sourceDirPath);
+
+        Files.walk(sourceDirAbsolutePath)
+            .forEach(
+                (sourcePath) -> {
+                  final @NotNull Path relativePath = sourceDirAbsolutePath.relativize(sourcePath);
+                  final @NotNull Path destinationPath = outputDir.toPath().resolve(relativePath);
+
+                  if (sourcePath.toFile().isFile()) {
+                    try {
+                      Files.createDirectories(destinationPath.getParent());
+                      Files.copy(sourcePath, destinationPath, StandardCopyOption.REPLACE_EXISTING);
+                    } catch (IOException e) {
+                      logger.error(
+                          "Failed to copy file from {} to {}", sourcePath, destinationPath, e);
+                    }
+                  }
+                });
```

Comment:
To be on the safe side, I think we should use try-with-resource for the `Files.walk` stream as per the [oracle docs](https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/nio/file/Files.html#walk(java.nio.file.Path,java.nio.file.FileVisitOption...))
```suggestion
  try (final @NotNull Stream<Path> stream = Files.walk(sourceDirAbsolutePath)) {
      stream.forEach(
          (sourcePath) -> {
              final @NotNull Path relativePath = sourceDirAbsolutePath.relativize(sourcePath);
              final @NotNull Path destinationPath = outputDir.toPath().resolve(relativePath);
  
              if (sourcePath.toFile().isFile()) {
                  try {
                      Files.createDirectories(destinationPath.getParent());
                      Files.copy(sourcePath, destinationPath, StandardCopyOption.REPLACE_EXISTING);
                  } catch (IOException e) {
                      logger.error(
                          "Failed to copy file from {} to {}", sourcePath, destinationPath, e);
                  }
              }
          }
      );
  }
```